### PR TITLE
box: add box.is_configured

### LINKED ISSUE #7255: Allow box.info before box.cfg
There is a local variable, that is not exposed:
https://github.com/tarantool/tarantool/blob/a353ec918756c889e2c98f62617050a574a0168e/src/box/lua/load_cfg.lua#L739

Currently, the only way to check is to do
```
type(box.cfg) == 'function'
```

Proposed change:
```
tarantool> box.info
---
- error: "builtin/box/load_cfg.lua:726: Please call box.cfg{} first\nstack traceback:\n\tbuiltin/box/load_cfg.lua:726:
    in function '__index'\n\t[string \"return box.info\"]:1: in main chunk\n\t[C]:
    in function 'pcall'\n\tbuiltin/box/console.lua:404: in function 'eval'\n\tbuiltin/box/console.lua:710:
    in function 'repl'\n\tbuiltin/box/console.lua:759: in function <builtin/box/console.lua:747>"
...

tarantool> box.is_configured
---
- false
...

tarantool> box.cfg{}
...skip...
---
...

tarantool> box.is_configured
---
- true
...
```

### PR DESCRIPTION
**Pull request by nshy**
It shows whether box.cfg() is called or not.

Closes #7255

@TarantoolBot document
Title: add box.is_configured

It shows whether box.cfg() is called or not.

### CODE REVIEW

**Comment by R-omk** [comment]
> It shows whether box.cfg() is called or not.

'called' does not always mean 'configured', maybe it is worth formulating the request for documentation in a different way ?

Maybe I'm wrong and it always happens atomic and simultaneously.

File: src/box/lua/load_cfg.lua [hunk 1]
```diff
@@ -749,13 +749,13 @@ setmetatable(box, {
 -- `true` when box is configured.
 --
 -- Use locked() wrapper to obtain reliable results.
-local box_is_configured = false
+box.is_configured = false
```
Comment (Totktonada) [inline src/box/lua/load_cfg.lua:752]: I would highlight the comment:

> Use locked() wrapper to obtain reliable results.

Why it gives a reliable result for a user if it doesn't for our code?

Comment (nshy) [reply src/box/lua/load_cfg.lua:752]: Ok, I rephrase to use 'configured', it is more precise when we have multiple fibers.

Comment (nshy) [reply src/box/lua/load_cfg.lua:752]: AFAIU comment for using locked() is misplaced, it should be placed before 'local function load_cfg(cfg)' which is just a few lines below.

Comment (Totktonada) [reply src/box/lua/load_cfg.lua:752]: AFAIR, no, it was mine comment and I have added it intentionally. However now I'm looking into the code (859df3a71b38c8d12047de06ed50e8c040fe2985) and don't see why not to access it without locks. Maybe it was overcaution.

Please, drop the comment. I'll look around a bit more and if I'll find what is not correct with the direct access, I'll let you know.

**Comment by Totktonada** [comment]
I'll cite @Mons's requirement from the issue:

> It is required to avoid hardly-interruptible exceptions of indexing box.
> I.e.: box_is_configured should have the meaning: I can freely access box[anything] and will not blow up on `Please call box.cfg{} first`

box configuration has several phases. For purposes of this discussion I'll point the following phases (it is a rough list):

1. Normalize and validate options.
2. Restore `box.<...>` members (except `execute`). Stop raising `Please call box.cfg{} first` errors on accessing the members.
3. Apply non-dynamic and some dynamic options, initialize subsystems.
4. Load the database (local or remote recovery). !! This process yields periodically. !!
5. At this point `box_is_configured()` from `box.cc` starts to return `true`.
6. Initialize RAFT subsystem.
7. Apply remaining dynamic options.
8. Uprade the database schema if needed and possible.
9. Restore `box.execute` (see e8d5515a688e60f56590869e6b69c57d59188124).
10. Set `box_is_configured` from `load_cfg.lua` to `true`.
11. Start `flightrec` if available.

Technically speaking the variable from 10th phase adheres @Mons's requirement. However I guess that he assumed 2nd phase (which, I guess, actually visible somewhere during 4th, at first yield).

If we'll expose a flag after the 2nd phase, we should reimplement `box.execute()` in the way that will move it from 9th to 2nd phase.

Also, if we'll do it, the database may be in a very initial state of loading. I'm not ever sure that system spaces are guaranteed to appear here.

I'm not sure that we should say something like 'box is more or less ready' at this stage. But I guess it is what the user want.

Let's discuss?

**Comment by Totktonada** [comment]
We discussed it with @Mons adn @nshy and decided to provide `box.info()` before first `box.cfg()` call instead of exposing the `box.is_configured` flag. Points to take care:

* On early configuration stages not all values are available. We can omit them until box will step into particular stage.
* There are values that are reported as `-1` or `box.NULL` on early configuratoin stages. Let's report them this way to be compatible with existing behaviour.

We also agreed that we can add information about current loading stage into `box.info()`, but it is better to do separately. I'll file an issue. Points to consider here:

* The rough list of staged above.
* More verbose information about stage 4 (recovery): whether all system spaces are available, whether we can use secondary indexes (they're build after all primary ones -- from Mons's words).
* It seems, we **can't** reuse `box.info().status` for this purpose due to backward compatibility concerns. Let's add another field.
* It would be nice to make the stages comparable. However it would be nice, at the same time, to make them huma readable. Raw idea: `'001 description'`, `'002 description'` and so on. Or, maybe, make them tables like `{num = <...>, name = <...>}`.

**Comment by Totktonada** [comment]
I moved the discussion about verbose box configuration stages into https://github.com/tarantool/tarantool/discussions/7563.

**Comment by nshy** [comment]
Discussed with @Totktonada 

Currently box.info() in parallel fiber during box.cfg{} shows the same result as call before box.cfg{} (if in whitelist) except for status ('loading' in this case). So we should keep these values for backcompat (can not change them to nil).

**Pull request closed without merging**